learning_journal/tests.py

```diff
@@ -1,65 +1,249 @@
+    req = dummy_request
+    req.matchdict = {"id": "3"}
+    req.method = "POST"
+    req.POST["title"] = "new thing"
+    req.POST["creation_date"] = datetime.datetime(2017, 2, 24, 0, 0)
+    req.POST["body"] = "and more new things"
+    try:
+        edit(req)
+    except:
+        new_title = dummy_request.dbsession.query(Entry).get(3).title
+        assert new_title == "new thing"
+
+
+def test_write_view_updates_db(dummy_request, add_models):
+    """Test that the write view adds entries to the database."""
+    from learning_journal.views.default import write
+    row_count_before_post = dummy_request.dbsession.query(Entry).count()
+    req = dummy_request
+    req.method = "POST"
+    req.POST["title"] = "a new post"
+    req.POST["title1"] = "a new subtitle"
+    req.POST["creation_date"] = datetime.datetime(2016, 12, 20, 0, 0)
+    req.POST["body"] = "a new body"
+    try:
+        write(req)
+    except:
+        row_count_after_post = dummy_request.dbsession.query(Entry).count()
+        assert row_count_after_post == row_count_before_post + 1
+
+
+
+#Functional test stuff.
+
+@pytest.fixture
+def testapp(request):
+    """Create an app instance for testing."""
+    from webtest import TestApp
+    from learning_journal import main
+    app = main({})
+    def main(global_config, **settings):
+        """Return a Pyramid WSGI application."""
+        config = Configurator(settings=settings)
+        config.include('pyramid_jinja2')
+        config.include('learning_journal.models')
+        config.include('learning_journal.routes')
+        config.include('learning_journal.security')
+        config.scan()
+        return config.make_wsgi_app()
+
+    testapp = TestApp(app)
+    session_factory = app.registry["dbsession_factory"]
+    session = session_factory()
+    engine = session.bind
+    Base.metadata.drop_all(engine)
+    Base.metadata.create_all(engine)
+
+    def tearDown():
+        Base.metadata.drop_all(engine)
+
+    request.addfinalizer(tearDown)
+    return testapp
+
+
+@pytest.fixture
+def fill_the_db(testapp):
+    """Fill database with journal entries."""
+    SessionFactory = testapp.app.registry["dbsession_factory"]
+    with transaction.manager:
+        dbsession = get_tm_session(SessionFactory, transaction.manager)
+        for entry in ENTRIES:
+            post = Entry(title=entry["title"], creation_date=entry["creation_date"], body=entry["body"])
+            dbsession.add(post)
+    return dbsession
+
+
+def test_non_authenticated_user_cannot_access_write_view(testapp):
+    """Test that accessing create new post is forbidden without auth."""
+    response = testapp.get('/journal/write', status=403)
+    assert response.status_code == 403
+
+def test_non_authenticated_user_cannot_access_edit_view(testapp):
+    """Test that accessing create new post is forbidden without auth."""
+    response = testapp.get('/journal/1/editentry', status=403)
+    assert response.status_code == 403
```

Comment:
You need more functional tests. Just generally. This barely tests all of 2 routes and you have several more than that.